internal/service/vmservice/vm.go

```diff
@@ -257,21 +259,70 @@ func getMachineAddresses(scope *scope.MachineScope) ([]clusterv1.MachineAddress,
 		},
 	}
 
+	networkSpec := scope.ProxmoxMachine.Spec.Network
+	if networkSpec == nil {
+		networkSpec = &infrav1alpha1.NetworkSpec{
+			Default: &infrav1alpha1.NetworkDevice{},
+		}
+	}
+	var ipv4, ipv6 string
+	// get ip addresses from vm if dhcp is enabled
+	if hasDHCPEnabled(scope.InfraCluster.ProxmoxCluster.Spec.IPv4Config,
+		networkSpec.Default,
+		infrav1alpha1.DefaultNetworkDevice,
+		infrav1alpha1.IPV4Format) ||
+		hasDHCPEnabled(scope.InfraCluster.ProxmoxCluster.Spec.IPv6Config,
+			networkSpec.Default,
+			infrav1alpha1.DefaultNetworkDevice,
+			infrav1alpha1.IPV6Format) {
+		nets, err := scope.InfraCluster.ProxmoxClient.GetVMNetwork(ctx, scope.VirtualMachine)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to get network interfaces for vm %s", scope.ProxmoxMachine.GetName())
+		}
+
+		for _, net := range nets {
+			// default device
+			mac := extractMACAddress(scope.VirtualMachine.VirtualMachineConfig.Net0)
+			if strings.EqualFold(mac, net.HardwareAddress) {
+				for _, ip := range net.IPAddresses {
+					if ip.IPAddressType == "ipv4" && ip.IPAddress != scope.InfraCluster.ProxmoxCluster.Spec.ControlPlaneEndpoint.Host {
+						ipv4 = ip.IPAddress
+					} else if ip.IPAddressType == "ipv6" && !strings.Contains(ip.IPAddress, "fe80:") {
```

Comment:
Please use [`netip.Addr.IsLinkLocalUnicast`](https://pkg.go.dev/net/netip#Addr.IsLinkLocalUnicast). Substring match for IPv6 is always less accurate than comparing the bits.